### ISSUE TITLE
Proposal: bring ML into the rule language and add agent-ready telemetry

### ISSUE THREAD

**Issue by ssam18** [snort3/snort3#468]
Opening this as a proposal and discussion topic. Happy to be wrong about any of it, but wanted to put a structured pitch out so the team can react to something concrete instead of vibes.

## Background

Snort 3 already has LibML in the tree and the plugin model is flexible enough to host ML inspectors. What we don't have yet is a clear position on where ML fits in the detection stack. Right now it is a side door. You can register an ML inspector and have it emit its own alerts, but rule authors cannot reference a model output from inside a regular rule.

Meanwhile every commercial NDR vendor leads with "AI-powered" in their marketing, and Suricata is starting to see third-party ML plugins land. The longer the project waits to take a position, the harder it gets to claim the AI-native open source IDS slot.

This proposal is two batches of work. Batch 1 is foundational and changes how ML and Snort interact at the engine level. Batch 2 is analyst-facing tooling that becomes much easier to build once the foundation is in place.

A middle batch (rule authoring copilot, alert triage agent, adaptive inspection depth) is intentionally left out. All of that is straightforward follow on work once the foundation lands and does not need to be argued about now.

### 1. ML as a rule option

Add a new rule option, call it `ml_score:`, that lets a rule reference the output of a registered model the same way it references `content:` or `pcre:`. A rule could look like this.

```
alert tcp any any -> any 443 (
    msg:"suspicious TLS handshake";
    ja4:"t13d1516h2_...";
    ml_score:tls_malware_classifier, gt 0.85;
    sid:1000001; rev:1;
)
```

Under the hood we would need a model registry that loads at startup (ONNX or LibML native format), a base inspector class for ML models, and a rule option evaluator that pulls the score out and feeds it to the rule engine.

**Why it matters?** Today an ML inspector either emits its own alerts or dumps to a side channel. Rule authors cannot combine model output with the rest of the rule language. They cannot say "match this regex AND score above threshold on the malware classifier." Hybrid rules of that shape consistently outperform either pure signature or pure ML in published NIDS literature. It also keeps ML-influenced detections auditable the same way every other rule is, by reading the rule.

**Risks:** Latency is the obvious one. Inline IPS cannot afford a 200µs model call on every packet. Three mitigations would address most of it. First, model evaluation only fires after cheaper rule options have narrowed the candidate set. Second, an async evaluation mode for IDS-only deployments. Third, per-model latency budgets with a circuit breaker if a model goes over.

### 2. Encrypted traffic classification using JA4 plus small models

Ship a bundle of small pretrained models (JA4, JA4S, JA4H plus a handful of behavioral features like packet timing and byte ratios) that classify TLS flows into known good, known malware family, or anomalous. Distribute them through the same channel as rules.

**Why it matters?** TLS 1.3, ECH, and QUIC together have made the encryption problem for signature based IDS noticeably worse. Exact JA4 fingerprint matching closes part of the gap but the long tail of variants needs a model. This is probably the biggest detection coverage hole that Snort can close in the next year.

Talos has a training data advantage that no Suricata vendor can match at scale. That is a real advantage and hard to copy regardless of who replicates the engine work. The feature also lands cleanly on top of item 1. Ship the models, expose them through `ml_score:`, distribute rules that use them.

**Risks:** Adversarial drift. Once attackers see what we flag they reshape their stacks. Mitigations are an aggressive refresh cadence (weeks not months), calibrated confidence outputs so analysts can tune thresholds, and a clear retirement path for stale models.

### 3. Telemetry an agent can consume

Extend the output side so Snort emits an event stream designed for downstream LLM-based tools. Specifically.

- Per alert MITRE ATT&CK technique mapping where the rule metadata supports it
- Stable correlation IDs linking related events (same flow, same JA4, same host pair within a time window)
- Optional embedding vectors on flow records, for similarity search downstream
- A documented schema and stable event ontology, with a real versioning story

This is not "add another logger." It is deliberate schema work, comparable in scope to what EVE was for Suricata in the JSON era.

**Why it matters?** EVE is the reason every SIEM and SOAR vendor in the last decade integrated Suricata cleanly. The equivalent for the agent era is a telemetry contract that LLM-based tools can consume directly without building their own enrichment layer on top. Snort's logging today is fine for humans and for SIEMs but does not anticipate agents at all.

This is also the lowest risk item in batch 1. No novel ML, no inline latency concerns. Schema and plumbing.

**Risks:** Schema design done badly is worse than no schema. You end up with a half committed contract nobody trusts. This needs a small focused design group early and a public RFC process so external integrators can comment before anything is frozen.

## Batch 2: analyst tooling

These sit on top of the telemetry stream from item 3. They do not need to live inside the Snort engine. They can ship as companion projects under the Snort organization. Including them here so the strategic value of batch 1 is visible to users, not just to engine maintainers.

### 4. Hunting copilot

A chat interface over the Snort event stream. Analyst asks "show me hosts that talked to a newly registered domain and then transferred more than 50MB in the next hour." The LLM translates that to a query, runs it over the telemetry store, summarizes the result.

The floor for good threat hunting today is an analyst who knows the schema cold and writes good queries. That is a small population. Lowering that floor is the biggest productivity win the project can offer SOC teams, and it costs almost no engineering once the schema from item 3 is stable.

### 5. Session embeddings with nearest neighbor lookup

Compute embeddings on TCP sessions, HTTP transactions, TLS handshakes. Index them. For new flows, find their nearest neighbors in the known malicious corpus and score based on closeness.

This catches malware variants that signatures miss. Same family, different packing, different C2 domain. It also gives analysts a much better explanation than a binary classifier. "This flow is 0.94 cosine similar to a Cobalt Strike beacon variant seen three days ago in another environment" is a sentence an analyst can act on immediately. Suricata has a hard time doing this because they do not have the corpus or pipeline. Snort does.

### 6. Multi step attack reconstruction

An agent that stitches related alerts into kill chain narratives mapped to ATT&CK. Initial access via web exploit, persistence via scheduled task, lateral movement over SMB, and so on. Output is a structured incident object, not a stream of individual alerts.

This is the work that tier 2 and tier 3 analysts do all day. Automating it does not replace them. It lets them spend their time on response and hunting instead of correlation. Once batch 1 item 3 and batch 2 item 5 exist, this is mostly a graph construction and prompting problem, not a research problem.

## Suggested order

Roughly this.

1. RFC for the telemetry schema (batch 1 item 3). Long pole because it needs community buy in, but it has no implementation dependencies, so start it first.
2. Engine and rule API for `ml_score:` (batch 1 item 1). Can land before any specific model exists. Largely parallel with the RFC.
3. First production model bundle (batch 1 item 2). Depends on the rule option landing.
4. Begin batch 2 work as companion projects, picking based on community pull.

## What I am asking from the team

Three things, in order of importance.

1. A direction call on whether ML belongs in the rule language at all, or whether it should remain a side channel feature. That decision gates whether batch 1 item 1 is in scope.

2. A small design group for the telemetry RFC. Two or three people from the core team plus an open invitation to external integrators. This contract needs to land cleanly the first time.

3. Guidance on how model bundles get distributed. Through the rules update channel, a separate channel, or via the plugin loader. That choice affects the operational story for batch 1 item 2.

If the direction call is yes, I am happy to take a first pass at the telemetry RFC and a prototype of the `ml_score:` rule option as a starting point.
